actions/concurrent_runner.go

```diff
@@ -0,0 +1,100 @@
+// pmm-agent
+// Copyright (C) 2018 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package actions
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+const defaultTimeout = time.Second * 10
+
+// ActionResult represents an action result.
+//nolint:unused
+type ActionResult struct {
+	ID             string
+	Name           string
+	Error          error
+	CombinedOutput []byte
+}
+
+// ConcurrentRunner represents concurrent action runner.
+// Action runner is component that can run an Actions.
+//nolint:unused
+type ConcurrentRunner struct {
+	out    chan ActionResult
+	logger logrus.FieldLogger
+
+	actions sync.Map // map[string]Action
```

Comment:
`sync.Map` for special cases only. It is not a generic-purpose map+Lock replacement. https://www.youtube.com/watch?v=C1EtfDnsdDs

Reply:
https://golang.org/pkg/sync/#Map
> The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or **(2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys.** In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex.

It's **2** isn't it? It uses inside multiple goroutines, and it read/write disjoint sets of keys.